File: killer_sudoku/solver/engine/backtracker.py
"""MRV backtracker fallback for the solver engine.

Called when the rule-based engine stalls and cannot make further progress.
Applies MRV (Minimum Remaining Values) cell ordering with forward checking
(peer-elimination propagation + naked-single cascade) to efficiently search
for a solution.

Supports both classic sudoku (row/col/box constraints only) and killer sudoku
(row/col/box + cage sum constraints).
"""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

from killer_sudoku.solver.engine.board_state import BoardState

Cell = tuple[int, int]

# Precomputed peer sets: peers[r][c] = frozenset of all cells sharing a row,
# column, or 3x3 box with (r, c), excluding (r, c) itself.
# This is pure constant data computed once at import time.
_PEERS: list[list[frozenset[Cell]]] = [
    [
        frozenset(
            {(r, c2) for c2 in range(9) if c2 != c}
            | {(r2, c) for r2 in range(9) if r2 != r}
            | {
                ((r // 3) * 3 + dr, (c // 3) * 3 + dc)
                for dr in range(3)
                for dc in range(3)
                if ((r // 3) * 3 + dr, (c // 3) * 3 + dc) != (r, c)
            }
        )
        for c in range(9)
    ]
    for r in range(9)
]


def mrv_backtrack(board: BoardState) -> npt.NDArray[np.intp] | None:
    """Find a solution via MRV backtracking from a partially-solved BoardState.

    Extracts cage constraints from the board spec, copies the current candidate
    sets, and searches for a valid completion.  Forward checking (peer-constraint
    propagation + naked-single cascade on each assignment) keeps the branching
    factor small.

    Args:
        board: A BoardState that the rule engine was unable to fully solve.

    Returns:
        A (9, 9) array of digits 1-9 for the unique solution, or None if no
        valid completion exists (puzzle unsolvable from this candidate state).
    """
    cage_of = [[0] * 9 for _ in range(9)]
    cage_total: dict[int, int] = {}
    cage_cells: dict[int, list[Cell]] = {}

    for r in range(9):
        for c in range(9):
            cid = int(board.spec.regions[r, c])  # 1-based cage id
            cage_of[r][c] = cid
            cage_cells.setdefault(cid, []).append((r, c))
            t = int(board.spec.cage_totals[r, c])
            if t != 0:
                cage_total[cid] = t

    cands: list[list[set[int]]] = [
        [set(board.candidates[r][c]) for c in range(9)] for r in range(9)
    ]

    grid = _search(cands, cage_of, cage_total, cage_cells)
    if grid is None:
        return None
    return np.array(grid, dtype=np.intp)


def _peers_of(r: int, c: int) -> frozenset[Cell]:
    return _PEERS[r][c]


def _cage_valid(
    cands: list[list[set[int]]],
    cid: int,
    cage_total: dict[int, int],
    cage_cells: dict[int, list[Cell]],
) -> bool:
    """Return False if the cage constraint is provably violated."""
    total = cage_total.get(cid)
    if total is None:
        return True

    cells = cage_cells[cid]
    placed_sum = 0
    remaining = 0
    for r2, c2 in cells:
        s = cands[r2][c2]
        if not s:
            return False
        if len(s) == 1:
            placed_sum += next(iter(s))
        else:
            remaining += 1

    if placed_sum > total:
        return False
    if remaining == 0:
        return placed_sum == total

    # Check that the remaining cells could still sum to (total - placed_sum).
    needed = total - placed_sum
    min_fill = sum(range(1, remaining + 1))  # 1 + 2 + ... + remaining
    max_fill = sum(range(10 - remaining, 10))  # (10-remaining) + ... + 9
    return min_fill <= needed <= max_fill


def _assign(
    cands: list[list[set[int]]],
    r: int,
    c: int,
    d: int,
    cage_of: list[list[int]],
    cage_total: dict[int, int],
    cage_cells: dict[int, list[Cell]],
) -> bool:
    """Place digit d at (r, c), propagate, and validate.

    Runs peer-elimination with naked-single cascade until quiescent.
    Validates cage-sum constraints after each placement.

    Returns False on contradiction (empty candidate set or violated cage sum).
    """
    cands[r][c] = {d}
    queue: list[tuple[int, int, int]] = [(r, c, d)]

    while queue:
        r0, c0, d0 = queue.pop()

        # Validate the cage of the placed cell.
        if not _cage_valid(cands, cage_of[r0][c0], cage_total, cage_cells):
            return False

        # Eliminate d0 from all peers, cascade naked singles.
        for r2, c2 in _peers_of(r0, c0):
            s = cands[r2][c2]
            if d0 not in s:
                continue
            s.discard(d0)
            if not s:
                return False
            if len(s) == 1:
                d_new = next(iter(s))
                queue.append((r2, c2, d_new))
                # Validate the cage of the newly placed cell.
                if not _cage_valid(cands, cage_of[r2][c2], cage_total, cage_cells):
                    return False

    return True


def _search(
    cands: list[list[set[int]]],
    cage_of: list[list[int]],
    cage_total: dict[int, int],
    cage_cells: dict[int, list[Cell]],
) -> list[list[int]] | None:
    """Recursive MRV search.  Returns a solved 9x9 grid or None on failure."""
    # MRV: find the unsolved cell with fewest candidates.
    min_count = 10
    best: Cell | None = None
    for r in range(9):
        for c in range(9):
            n = len(cands[r][c])
            if n == 0:
                return None
            if 1 < n < min_count:
                min_count = n
                best = (r, c)

    if best is None:
        # Every cell is determined — puzzle solved.
        return [[next(iter(cands[r][c])) for c in range(9)] for r in range(9)]

    r, c = best
    for d in sorted(cands[r][c]):
        # Deep-copy candidate sets for this branch.
        new_cands: list[list[set[int]]] = [[set(s) for s in row] for row in cands]
        if _assign(new_cands, r, c, d, cage_of, cage_total, cage_cells):
            result = _search(new_cands, cage_of, cage_total, cage_cells)
            if result is not None:
                return result

    return None  # All candidates exhausted — backtrack.

File: web/src/engine/backtracker.ts
/**
 * MRV backtracker fallback for the solver engine.
 *
 * Mirrors Python's `killer_sudoku.solver.engine.backtracker` module.
 *
 * Called when the rule-based engine stalls. Applies MRV (Minimum Remaining
 * Values) cell ordering with forward checking (peer elimination + naked-single
 * cascade) to search for a solution.
 *
 * Works identically for classic and killer sudoku — the cage sum constraints
 * are included in the validity check when cage_total > 0.
 */

import type { BoardState } from './boardState.js';
import type { Cell } from './types.js';

// ---------------------------------------------------------------------------
// Precomputed peer sets: peers[r][c] = all cells sharing a row, col, or box
// ---------------------------------------------------------------------------

const PEERS: readonly (readonly Cell[])[][] = Array.from({length: 9}, (_, r) =>
  Array.from({length: 9}, (__, c) =>
    [
      ...Array.from({length: 9}, (_, c2): Cell => [r, c2] as Cell).filter(([, c2]) => c2 !== c),
      ...Array.from({length: 9}, (_, r2): Cell => [r2, c] as Cell).filter(([r2]) => r2 !== r),
      ...Array.from({length: 9}, (_, k): Cell => [
        (r / 3 | 0) * 3 + (k / 3 | 0),
        (c / 3 | 0) * 3 + (k % 3),
      ] as Cell).filter(([pr, pc]) => !(pr === r && pc === c)),
    ].filter(([pr, pc], i, arr) =>
      arr.findIndex(([qr, qc]) => qr === pr && qc === pc) === i
    )
  )
);

/**
 * Cage-sum data for mrvBacktrack's validity check — built by
 * KillerBoardState.cageConstraints(); plain BoardState has none (null).
 */
export interface CageConstraints {
  readonly cageOf: number[][];
  readonly cageTotal: ReadonlyMap<number, number>;
  readonly cageCells: ReadonlyMap<number, readonly Cell[]>;
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

/**
 * Find a solution via MRV backtracking from a partially-solved BoardState.
 *
 * Asks the board for its cage constraints (null for a plain classic board —
 * search() then degrades to pure row/col/box backtracking), copies current
 * candidate sets, and searches for a valid completion. Forward checking keeps
 * the branching factor small.
 *
 * Returns a 9×9 grid of placed digits, or null if unsolvable from this state.
 */
export function mrvBacktrack(board: BoardState): number[][] | null {
  const constraints = board.cageConstraints();
  const cageOf: number[][] = constraints?.cageOf ?? Array.from({length: 9}, () => new Array<number>(9).fill(0));
  const cageTotal: ReadonlyMap<number, number> = constraints?.cageTotal ?? new Map();
  const cageCells: ReadonlyMap<number, readonly Cell[]> = constraints?.cageCells ?? new Map();

  const cands: Set<number>[][] = Array.from({length: 9}, (_, r) =>
    Array.from({length: 9}, (__, c) => new Set(board.cands(r, c))));

  const solution = search(cands, cageOf, cageTotal, cageCells, { n: 0 });
  if (solution !== null && !gridValid(solution)) {
    console.error('mrvBacktrack: search returned an invalid solution — treating as unsolvable');
    return null;
  }
  return solution;
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Returns true iff the 9×9 grid has no duplicate in any row, column, or 3×3 box. */
function gridValid(grid: number[][]): boolean {
  for (let i = 0; i < 9; i++) {
    const rowSeen = new Set<number>(), colSeen = new Set<number>();
    for (let j = 0; j < 9; j++) {
      const rv = grid[i]![j]!;
      const cv = grid[j]![i]!;
      if (rowSeen.has(rv) || colSeen.has(cv)) return false;
      rowSeen.add(rv); colSeen.add(cv);
    }
  }
  for (let br = 0; br < 3; br++) {
    for (let bc = 0; bc < 3; bc++) {
      const boxSeen = new Set<number>();
      for (let r = br * 3; r < br * 3 + 3; r++) {
        for (let c = bc * 3; c < bc * 3 + 3; c++) {
          const v = grid[r]![c]!;
          if (boxSeen.has(v)) return false;
          boxSeen.add(v);
        }
      }
    }
  }
  return true;
}

/**
 * Check whether a cage's current candidate state is still consistent with its sum constraint.
 *
 * Sums placed digits (singleton candidate sets) and checks that the remaining unplaced
 * cells can still reach the target total using distinct digits 1–9. Returns true if no
 * constraint is violated and the sum remains reachable; returns false immediately if any
 * cell has an empty candidate set or the placed sum already exceeds the total.
 */
function cageValid(
  cands: Set<number>[][],
  cid: number,
  cageTotal: ReadonlyMap<number, number>,
  cageCells: ReadonlyMap<number, readonly Cell[]>,
): boolean {
  const total = cageTotal.get(cid);
  if (total === undefined) return true;
  const cells = cageCells.get(cid)!;
  let placedSum = 0, remaining = 0;
  for (const [r, c] of cells) {
    const s = cands[r]![c]!;
    if (s.size === 0) return false;
    if (s.size === 1) placedSum += s.values().next().value as number;
    else remaining++;
  }
  if (placedSum > total) return false;
  if (remaining === 0) return placedSum === total;
  const needed = total - placedSum;
  const minFill = (remaining * (remaining + 1)) >> 1;
  const maxFill = remaining * 9 - ((remaining * (remaining - 1)) >> 1);
  return minFill <= needed && needed <= maxFill;
}

/**
 * Assign digit `d` to cell (r, c) and propagate via unit-arc consistency.
 *
 * Mutates `cands` in place. Eliminates `d` from all peers (same row/col/box);
 * when a peer is reduced to a singleton, enqueues it for further propagation.
 * Validates the affected cage after each step. Returns false (contradiction) if
 * any cell's candidate set becomes empty or any cage constraint is violated;
 * otherwise returns true.
 */
function assign(
  cands: Set<number>[][],
  r: number,
  c: number,
  d: number,
  cageOf: number[][],
  cageTotal: ReadonlyMap<number, number>,
  cageCells: ReadonlyMap<number, readonly Cell[]>,
): boolean {
  cands[r]![c] = new Set([d]);
  const queue: Array<[number, number, number]> = [[r, c, d]];

  while (queue.length > 0) {
    const [r0, c0, d0] = queue.pop()!;
    if (!cageValid(cands, cageOf[r0]![c0]!, cageTotal, cageCells)) return false;
    for (const [r2, c2] of PEERS[r0]![c0]!) {
      const s = cands[r2]![c2]!;
      if (!s.has(d0)) continue;
      s.delete(d0);
      if (s.size === 0) return false;
      if (s.size === 1) {
        const dNew = s.values().next().value as number;
        queue.push([r2, c2, dNew]);
        if (!cageValid(cands, cageOf[r2]![c2]!, cageTotal, cageCells)) return false;
      }
    }
  }
  return true;
}

const MAX_BACKTRACK_NODES = 100_000;

/**
 * Recursive MRV backtracking search over candidate sets.
 *
 * Selects the unassigned cell with the fewest remaining candidates (minimum remaining
 * values heuristic), tries each candidate via `assign`, and recurses. Returns the
 * completed 9×9 grid on success or `null` on contradiction or node-limit breach
 * (`counter.n > MAX_BACKTRACK_NODES`).
 */
function search(
  cands: Set<number>[][],
  cageOf: number[][],
  cageTotal: ReadonlyMap<number, number>,
  cageCells: ReadonlyMap<number, readonly Cell[]>,
  counter: { n: number },
): number[][] | null {
  if (++counter.n > MAX_BACKTRACK_NODES) return null;

  let minCount = 10;
  let best: [number, number] | null = null;
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      const n = cands[r]![c]!.size;
      if (n === 0) return null;
      if (n > 1 && n < minCount) { minCount = n; best = [r, c]; }
    }
  }
  if (best === null) {
    // All cells are singletons. Validate rows, columns, and boxes before returning —
    // contradictory board states can leave a wrong singleton due to applyEliminations'
    // size-≤1 guard, producing an all-singleton grid that is not a valid sudoku solution.
    for (let i = 0; i < 9; i++) {
      const rowSeen = new Set<number>(), colSeen = new Set<number>();
      for (let j = 0; j < 9; j++) {
        const rv = cands[i]![j]!.values().next().value as number;
        const cv = cands[j]![i]!.values().next().value as number;
        if (rowSeen.has(rv) || colSeen.has(cv)) return null;
        rowSeen.add(rv); colSeen.add(cv);
      }
    }
    for (let br = 0; br < 3; br++) {
      for (let bc = 0; bc < 3; bc++) {
        const boxSeen = new Set<number>();
        for (let r = br * 3; r < br * 3 + 3; r++) {
          for (let c = bc * 3; c < bc * 3 + 3; c++) {
            const v = cands[r]![c]!.values().next().value as number;
            if (boxSeen.has(v)) return null;
            boxSeen.add(v);
          }
        }
      }
    }
    return Array.from({length: 9}, (_, r) =>
      Array.from({length: 9}, (__, c) => cands[r]![c]!.values().next().value as number));
  }

  const [r, c] = best;
  for (const d of [...cands[r]![c]!].sort((a, b) => a - b)) {
    const newCands: Set<number>[][] = Array.from({length: 9}, (_, r2) =>
      Array.from({length: 9}, (__, c2) => new Set(cands[r2]![c2]!)));
    if (assign(newCands, r, c, d, cageOf, cageTotal, cageCells)) {
      const result = search(newCands, cageOf, cageTotal, cageCells, counter);
      if (result !== null) return result;
    }
  }
  return null;
}
